text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+  stage limit that CodePipeline has. With the new implementation, the actions will automatically spread over multiple
+  stages, even if you deploy multiple applications in parallel.
+- For simple use cases, you no longer need to manage CodePipeline Artifacts: artifact management is implicit
+  as objects that produce artifacts can be used to reference them.
+
+Best of all: your old code--even though it uses no longer recommended idioms--still continues to work, so you
+can upgrade to the new style at your own pace.
+
+The following:
+
+```ts
+const sourceArtifact = new codepipeline.Artifact();
+const cloudAssemblyArtifact = new codepipeline.Artifact('CloudAsm');
+const integTestArtifact = new codepipeline.Artifact('IntegTests');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  cloudAssemblyArtifact,
+
+  // Where the source can be found
+  sourceAction: new codepipeline_actions.GitHubSourceAction({
+    actionName: 'GitHub',
+    output: sourceArtifact,
+    oauthToken: SecretValue.secretsManager('github-token'),
+    owner: 'OWNER'
+    repo: 'REPO',
+    trigger: codepipeline_actions.GitHubTrigger.POLL,
+  }),
+
+  // How it will be built
+  synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
+    sourceArtifact,
+    cloudAssemblyArtifact,
+    projectName: 'MyServicePipeline-synth',
+    additionalArtifacts: [
+      {
+        directory: 'test',
+        artifact: integTestArtifact,
+      },
+    ],
+  }),
+});
+
+const stage = pipeline.addApplicationStage(new MyStage(this, 'PreProd', {
+  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
+}));
+stage.addActions(
+  new cdkp.ShellScriptAction({
+    actionName: 'UseSource',
+    commands: [
+      // Comes from source
+      'cat README.md',
+    ],
+    additionalArtifacts: [sourceArtifact],
+  }),
+);
+```
+
+Becomes:
+
+```ts
+const source = cdkp.Source.gitHub('OWNER/REPO');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  synthAction: cdkp.Build.standardNpmSynth({
+    input: source,
+    projectName: 'MyServicePipeline-synth',
+    additionalOutputs: [
+      { directory: 'test', name: 'tests' },
+    ],
+  }),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
+});
+pipeline.addApplicationStage(stage, {
+  approvals: [
+    cdkp.Approval.shellScript({
+      input: source,
+      commands: ['cat README.md'],
+    }),
+  ],
+});
+```
+
+The biggest changes are: we are layering our own classes over the primitive CodePipeline API
+(`Source.gitHub` instead of `codepipeline_actions.GitHubSourceAction`) with improved ergonomics
+and defaults, and we are making artifacts implicit.
+
+> The biggest change we want to make but don't have an answer for at time of this writing is how to give users the
+> ability to tweak every individual part of every indiviual CodeBuild project that gets generated.
```

Comment:
I'm not convinced escape hatches are a sufficient solution. 

It also depends on what you mean by "escape hatch". If you mean fall down to L1 and we don't expose anything, it will look like:

```ts
pipeline.node.tryFindChild('Pipeline/CodeBuild10dbaa7/Resource').addPropertyOverride(...);
```

Which is... possible but not great.

If we want to expose L2s, it will look like:

```ts
pipeline.assets[3].codeBuild.addToRolePolicy(...);
```

Which is better, but presumes the `Pipeline` construct knows to expose AWS-specific L2s, which means it's already backend-specific, so not an "escape hatch" in that sense.

Also most of our L2s aren't mutable, so any form of "escape  hatching" will necessarily force people down to the L1 level.

Practical problems like "I want all CodeBuild projects to run in a given VPC" don't have a practical answer using escape hatches. There are too many properties that all need to be changed in tandem, with an awkward interface, for a request which is reasonable, obvious and common.

Reply:
I agree that "I want all CodeBuild projects should run in a VPC" is not something that should be solved via escape hatches. It should be a feature of the CodePipeline backend exposed through some abstract interface in the "top layer".

```ts
const codePipelines = new cdkp.CodePipelinesBackend(this, 'CodePipelineBackend', {
  runAllCodeBuildProjectsInAVpc: vpc,
  someOtherCodeSuiteSettings: 'foo'
});

new CdkPipeline(this, 'Pipeline', {
  backend: codePipelines,
  // ...
});
```

Reply:
As for escape hatches - I meant the first version. Escape hatches are by definition untyped. Otherwise, they are just part of the API :-). They should be left for last resort of course. I think a combination of abstract interfaces at the "top layer" and escape hatches "from the bottom" should be sufficient for 1.0

Reply:
I agree, it should work like this. The question is whether that is the API we want to present to users. For users just interested in CodePipeline, they are now paying a tax to enable use cases they're not interested in.

We can remove that tax again by adding a specialized class like:

```ts
interface CodePipelinePipelineProps {
    runAllCodeBuildProjectsInAVpc: IVpc,
    someOtherCodeSuiteSettings: stirng;
}

class CodePipelinePipeline {
  constructor(scope, id, props) {
    const pipeline = new GenericPipeline(this, 'Resource', {
      backend: new CodePipelineBackend(props)
    });
  }
}
```

But until we have another backend tested and implemented and are ready to advertise (which we are definitely not going to do in the current time line), it's probably safer to just present the above `CodePipelinePipeline` class as the primary interface, have it implemented in terms of the more generic `CdkPipeline`, but hide the internals until we're more sure about them. The fundamental parts will there to pick up the work later on when we'll have time to validate them.